op-challenger/game/fault/trace/split/provider.go

```diff
@@ -11,39 +13,89 @@ import (
 
 var _ types.TraceProvider = (*SplitTraceProvider)(nil)
 
+// CreateLowerProvider creates a new [types.TraceProvider] that is
+// used for the bottom half of the [SplitTraceProvider].
+type CreateLowerProvider func(ctx context.Context, preClaim common.Hash, postClaim common.Hash) (types.TraceProvider, error)
+
+// AbsolutePrestateCommitmentFetcher is a function pointer to fetch the absolute prestate commitment
+type AbsolutePrestateCommitmentFetcher func(ctx context.Context) (hash common.Hash, err error)
+
+// AbsolutePrestateFetcher is a function pointer to fetch the absolute prestate
+type AbsolutePrestateFetcher func(ctx context.Context) (preimage []byte, err error)
+
 // SplitTraceProvider is a [types.TraceProvider] implementation that
 // routes requests to the correct internal trace provider based on the
 // depth of the requested trace.
 type SplitTraceProvider struct {
-	logger         log.Logger
-	topProvider    types.TraceProvider
-	bottomProvider types.TraceProvider
-	topDepth       uint64
+	logger                            log.Logger
+	upper                             types.TraceProvider
+	lower                             map[*big.Int]types.TraceProvider
+	lowerCreator                      CreateLowerProvider
+	absolutePrestateCommitmentFetcher AbsolutePrestateCommitmentFetcher
+	absolutePrestateFetcher           AbsolutePrestateFetcher
+	topDepth                          uint64
 }
 
 // NewTraceProvider creates a new [SplitTraceProvider] instance.
 // The [topDepth] parameter specifies the depth at which the internal
 // [types.TraceProvider] should be switched.
-func NewTraceProvider(logger log.Logger, topProvider types.TraceProvider, bottomProvider types.TraceProvider, topDepth uint64) *SplitTraceProvider {
+func NewTraceProvider(logger log.Logger, upper types.TraceProvider, creator CreateLowerProvider, topDepth uint64) *SplitTraceProvider {
 	return &SplitTraceProvider{
-		logger:         logger,
-		topProvider:    topProvider,
-		bottomProvider: bottomProvider,
-		topDepth:       topDepth,
+		logger:       logger,
+		upper:        upper,
+		lower:        make(map[*big.Int]types.TraceProvider),
+		lowerCreator: creator,
+		topDepth:     topDepth,
 	}
 }
 
-func (s *SplitTraceProvider) providerForDepth(depth uint64) (uint64, types.TraceProvider) {
-	if depth <= s.topDepth {
-		return 0, s.topProvider
+func (s *SplitTraceProvider) buildLowerProvider(ctx context.Context, pos types.Position) (types.TraceProvider, error) {
+	ancestorIndex := pos.IndexForParentDepth(int(s.topDepth))
+	provider, ok := s.lower[ancestorIndex]
```

Comment:
This won't be unique enough because there may be two claims at the same upper-leaf-node position with different values and they will need to have different lower trace providers (see other comment about using the pre/post claim from the game, not from our trace provider).

We could use the contract index of the leaf claim involved. Otherwise we'd need to use the pre and post claim values as the key.